### ISSUE TITLE
Allow assertions in instantiation tests

### ISSUE THREAD

**Issue by benshayden** [catapult-project/catapult#3026]
The [style guide](https://github.com/catapult-project/catapult/blob/master/docs/style-guide.md#tests) currently discourages assertions in instantiation tests, though I've never understood why.

I have found it tremendously useful in [histogram_set_table_test](https://github.com/catapult-project/catapult/blob/master/tracing/tracing/value/ui/histogram_set_table_test.html) to both display histogram-set-tables with various inputs and make assertions about the outputs. The assertions have caught real bugs, as have I as I manually interact with the output.

This came up in a [recent codereview](https://codereview.chromium.org/2515563002/#msg4) with @Apeliotes 

I could see an argument for limiting the number of instantiation tests to a number that a human could conceivably manually verify without losing their grip on reality, but that seems orthogonal to whether the test makes assertions. It also seems difficult in practice to limit the number of instantiation tests for swiss-army-knives like histogram-set-table.

Any objections to removing the sentence discouraging assertions in instantiation tests?

@zeptonaut @eakuefner

**Comment by eakuefner** [comment]
I agree that instantiation tests should be allowed to make assertions, and I feel like we should maybe remove this section. In fact, I wasn't even aware of the prohibition imposed in the style guide, and have *recommended* that people make judicious assertions where reasonable.

I don't think the term "instantiation test" is well-defined for any of us, and I don't see the harm in allowing assertions in such tests. My best guess is that this rule might have been added as a way to control overtesting, but if anything, to me, it might encourage it, by causing people to add spurious tests just for the sake of making assertions.

**Comment by zeptonaut** [comment]
I think that one goal of this part of the style guide is to try and guarantee that we test behaviors independently ([as suggested by the Google unit testing best practices](http://bit.ly/2fZZFKX)). Something like this could is a great test:

```js
test('timelineView_removingProcessRemovesTrack', function() {
  var model = new Model(/* some model with two processes */);
  var timelineView = new TimelineView(model);
  htmlOutput.add(timelineView);

  assert.lengthOf(timelineView.tracks, 2);
  model.processes.pop();
  assert.lengthOf(timelineView.tracks, 1);
});
```

whereas with something like this, we're testing multiple behaviors in a single test, and we really should have written multiple tests:

```js
test('instantiate', function() {
  var model = new Model(/* some model with two processes */);
  var timelineView = new TimelineView(model);
  htmlOutput.add(timelineView);

  assert.lengthOf(timelineView.tracks, 2);
  model.processes.pop();
  assert.lengthOf(timelineView.tracks, 1);
  model.vsyncs = [0, 16, 32, 48];
  assert.lengthOf(timelineView.vsyncStripes, 3);
});
```

I wonder if what the "don't assert in instantiation test" rule was trying to convey was that, if you're asserting in an instantiation test, it's not really an instantiation test: it's just a test that happens to have HTML output. I've seen lots of instantiation tests that follow the kind of wandering assertion pattern shown in the second example, and that's not good for the codebase.

**Comment by benshayden** [comment]
:+1: to testing behaviors independently, though that seems to me to be compatible with making assertions in instantiation tests.

Can you say more about what an "instantiation test" is? I haven't heard the term in any other context, and it was never explained to me in a way that I could relate it to other concepts. Googling turns up some interesting things, though none of them are particularly enlightening. [It was just that when Harry tried to relate those concepts to any other concepts, he drew a blank.](http://hpmor.com/chapter/109)

The name seems to imply that it only tests that something can be instantiated, which seems orthogonal to whether it has HTML output. If it only tests that something can be instantiated, then it seems to me like it might as well go further and assert something about the instantiation process, or else some functionality of the instantiated object, as long as it only tests one behavior, as you say.

Oh, unless "instantiation" is a kind of behavior, separate from anything else that the module might do? But how can a module perform another behavior if it is not first instantiated? Tests for non-instantiation behaviors would first need to instantiate the module. If instantiating a module fails, then multiple or perhaps all tests for that module would fail, violating the test independence clause. Instantiation tests themselves would seem to preclude testing a single behavior at a time.

It still feels like there's something I'm missing. :-)

**Comment by zeptonaut** [comment]
I've always understood instantiation tests to have two purposes:

1. Allow future people who are going to change that component to have a testbed to verify those changes didn't break anything (not useful for automated regression testing)
2. Act as a sort of `assert.doesNotThrow` for the behavior in the test.

**Comment by eakuefner** [comment]
I'm still a little unclear on what you mean by 2. Maybe some examples would help? Unit tests should not throw by definition, right? Because if they do, they fail.

**Comment by zeptonaut** [comment]
Also, just to be clear: I'm not 100% clear on what an instantiation test is either. I'm guessing that the main purpose of an instantiation test is just to provide playground for future people who might like to change the UI component to see the effects of those changes, and that the reason that they shouldn't make assertions is that there should be other, non-instantiation tests that programmatically ensure that the actual behaviors are correct.

The practical implications of this are that if your main goal is to verify a behavior, you should do the absolute minimum in that test to replicate the behavior and then assert that things happened as expected. Minimalism is key.

For an instantiation test, where the main goal is to provide a useful testbed rather than do the absolute minimum to verify a behavior, you might add a fair chunk of extra data so that the person viewing the testbed can see how the component reacts to having a good chunk of data.

As a practical example, if we were writing tests for a table element, we might have:

- A first test case that asserts that, when you add a row, the number of rows changes and that the first row contains the expected information
- A second test case that asserts that, after you add a row, you can remove it, and the number of rows decreases to zero
- A third test case that asserts that, when you add two rows, each contains the expected information
- An instantiation case where you add five rows to the table

Note, I have no idea if this is true, but it's consistent with many instantiation tests that I've seen in the code and seems to make sense to me.
